### PR TITLE
Add game rules for stability and influence

### PR DESCRIPTION
**Pull request by MichaelCourtney**
W.I.P.

### CODE REVIEW

**Comment by MichaelCourtney** [comment]
https://www.freeorion.org/forum/viewtopic.php?f=6&t=12214

**Comment by o01eg** [comment]
Do we need named value same as game rule? Maybe we can reference game rule instead of named value?

**Comment by MichaelCourtney** [comment]
> Do we need named value same as game rule? Maybe we can reference game rule instead of named value?

Presumably I can remove the named value and just have the game rule. I'll clean it up a little at the end

**Comment by MichaelCourtney** [comment]
Not certain whether the "Fixme" on the accounting label is correct. It was copied from another accounting label

File: default/scripting/game_rules.focs.py [hunk 1]
```diff
@@ -146,13 +146,73 @@
 GameRule(
     name="RULE_BASELINE_PLANET_STABILITY",
     description="RULE_BASELINE_PLANET_STABILITY_DESC",
-    category="BALANCE",
+    category="GOVERNMENT",
```
Comment (geoffthemedio) [inline default/scripting/game_rules.focs.py]: why a separate category for GOVERNMENT and not just BALANCE ?

**Comment by MichaelCourtney** [comment]
Length of option menu and scrollbar. It can be organised differently if you prefer.

**Comment by geoffthemedio** [comment]
I'd perhaps have a few "Balance: X" categories. Planet size could also be prefixed with "Balance: ". The existing Balance category could be "Balance: Misc" and/or maybe also have "Balance: Cost Scaling" split out.

Also, everything added is a stability modifier, I think? So perhaps call the category "Stability", not "Government" ?

**Comment by MichaelCourtney** [comment]
> I'd perhaps have a few "Balance: X" categories. Planet size could also be prefixed with "Balance: ". The existing Balance category could be "Balance: Misc" and/or maybe also have "Balance: Cost Scaling" split out.
> 
> Also, everything added is a stability modifier, I think? So perhaps call the category "Stability", not "Government" ?

There's 1 influence option for the capital. I'd have to see in-game. I'll test, the possible issue is Balance: Influence/Stability is truncated into oblivion, or the tab for Balance: Influence/Stability is too long. Will screenshot

**Comment by MichaelCourtney** [comment]
It creates large tabs. Up to you
Largetab

**Comment by MichaelCourtney** [comment]
Finished except for any clean up and tweaking

**Comment by geoffthemedio** [comment]
Maybe just shorten to "Balance:Planets" and "Balance:Stability"

**Comment by geoffthemedio** [comment]
@o01eg any idea what the stringtable linter error `default/stringtables/en.txt was reformatted` means?

**Comment by o01eg** [comment]
@geoffthemedio @MichaelCourtney  It propose next change:
```diff
@@ -275,6 +275,7 @@ Multiplayer
 BALANCE_STABILITY
 Balance:Stability
 
+
 ##
 ## Major errors
 ##
```

**Comment by geoffthemedio** [comment]
where does it say that?

**Comment by o01eg** [comment]
@geoffthemedio I called `./check/st-tool.py format default/stringtables/en.txt` manually and check what changed with `git diff`.

**Review by o01eg — APPROVED** [review]
Looks fine

**Comment by geoffthemedio** [comment]
my build might be messed up, but with this, the rightmost tab is an error string

**Comment by MichaelCourtney** [comment]
> my build might be messed up, but with this, the rightmost tab is an error string

I'll test it after work. Are you looking at it in English? The other stringtables will be missing entries, I'm not certain what to do about that.

**Comment by MichaelCourtney** [comment]
Ok it's working again. Though the order is a little jumbled. I'm not sure how it determines order. There were 2 problems:
options
The above are not defined in game_rules.focs.py, I actually couldn't find where they are defined.

The values were type=int when they actually needed to be signed integers. I can't see signed int anywhere on the other game rules so I've changed them back to float for now

**Comment by MichaelCourtney** [comment]
Sigh... Stuffed that up and had to revert and fix it again. Tried to squish the fix for 5 commits into the 1st one by accident...
Anyway... Should work outside of them appearing in an odd order in the options tab

**Comment by geoffthemedio** [comment]
> The values were type=int when they actually needed to be signed integers. 

I'm not sure what distinction you're making between "int" and "signed integers".

> I can't see signed int anywhere on the other game rules so I've changed them back to float for now

Some rules are defined in C++ when they are also used in C++ eg. https://github.com/freeorion/freeorion/blob/master/universe/Universe.cpp#L68

**Comment by MichaelCourtney** [comment]
> > The values were type=int when they actually needed to be signed integers.
> 
> I'm not sure what distinction you're making between "int" and "signed integers".

I'm saying that if I change them from type=float to type=int they all disappear from the options tab due to the possible negative values

**Comment by geoffthemedio** [comment]
Did you just change the `float` to `int` or did you also update the default, min, and max values to remove the `.0` part? If you pass it a value with a fractional part, it probably won't accept it for an int-valued field.

**Comment by MichaelCourtney** [comment]
> Did you just change the `float` to `int` or did you also update the default, min, and max values to remove the `.0` part? If you pass it a value with a fractional part, it probably won't accept it for an int-valued field.

I just changed them to int. You are right, I didn't consider the decimal. I'll try it out

**Comment by MichaelCourtney** [comment]
yeah that worked, will change them to int and update

**Comment by geoffthemedio** [comment]
This appears to have somewhat/intermittently broken the modified rules. From the unit tests of Ubuntu 20.04 on latest master:
````
Error: 6:27.517743 {0x00007f54bb649c00} [error] ai : GameRules.h:93 : bad any cast converting value of game rule named: RULE_BASELINE_PLANET_STABILITY. Returning default value instead
Error: 6:27.518009 {0x00007f54bb649c00} [error] ai : GameRules.h:97 : bad any cast converting default value of game rule named: RULE_BASELINE_PLANET_STABILITY. Returning data-type default value instead: 0
````

I'm not sure where this is looking up the wrong type... It shouldn't be in ValueRef GameRule lookups, as those will automatically handle casting. The issue also is sometimes not in the logs of the same test for different runs... Also not sure why about that.

**Comment by Grummel7** [comment]
> I'm not sure where this is looking up the wrong type... It shouldn't be in ValueRef GameRule lookups, as those will automatically handle casting. The issue also is sometimes not in the logs of the same test for different runs... Also not sure why about that.

This is strange. I wondered if it was related to #4124. When I looked for these errors, I saw them once, but then never again. The AI however consistently gets only default values, so most likely the two issues are not related.

**Pull request merged**